File: R/metrics.R
# Network metrics (Rust-backed) ----------------------------------------------

#' Basic street-network statistics
#'
#' Summary measures for an `osm_graph`: node and edge counts, total and mean
#' edge length, mean out-degree and self-loop count. Computation is performed
#' by the bundled Rust core.
#'
#' @param g An [osm_graph][new_osm_graph].
#' @param weight Edge column used as length/weight. Default `"length"`.
#'
#' @return A one-row [tibble][tibble::tibble] of statistics.
#' @export
#'
#' @examples
#' g <- example_osm_graph()
#' ox_basic_stats(g)
ox_basic_stats <- function(g, weight = "length") {
  stopifnot(is_osm_graph(g))
  ea <- graph_edge_arrays(g, weight)
  s <- rs_basic_stats(ea$from, ea$to, ea$weight, ea$n_nodes)
  tibble::as_tibble(s)
}

#' Street-orientation entropy
#'
#' Shannon entropy (in nats) of the distribution of edge compass bearings,
#' binned into equal sectors. Higher values indicate a more disordered
#' (organic) network; lower values a more ordered (gridiron) one.
#'
#' @param g An [osm_graph][new_osm_graph].
#' @param num_bins Number of equal bearing sectors over `[0, 360)`. Default `36`.
#'
#' @return A numeric scalar (entropy in nats).
#' @export
#'
#' @examples
#' g <- example_osm_graph()
#' ox_orientation_entropy(g)
ox_orientation_entropy <- function(g, num_bins = 36) {
  stopifnot(is_osm_graph(g))
  b <- ox_bearings(g)
  rs_orientation_entropy(b, as.integer(num_bins))
}

#' Compute edge compass bearings
#'
#' Initial compass bearing (degrees clockwise from north) of each edge, from
#' its first to its last coordinate. Geographic coordinates are used; projected
#' graphs are transformed to EPSG:4326 first.
#'
#' @param g An [osm_graph][new_osm_graph].
#'
#' @return A numeric vector of bearings, one per edge.
#' @export
#'
#' @examples
#' g <- example_osm_graph()
#' head(ox_bearings(g))
ox_bearings <- function(g) {
  stopifnot(is_osm_graph(g))
  edges <- g$edges
  if (!is.na(sf::st_crs(edges)) && sf::st_crs(edges) != sf::st_crs(4326)) {
    edges <- sf::st_transform(edges, 4326)
  }
  ends <- lapply(sf::st_geometry(edges), function(ls) {
    m <- sf::st_coordinates(ls)
    c(m[1, "X"], m[1, "Y"], m[nrow(m), "X"], m[nrow(m), "Y"])
  })
  ends <- do.call(rbind, ends)
  rs_bearings(ends[, 2], ends[, 1], ends[, 4], ends[, 3])
}

File: src/rust/src/metrics.rs
//! Urban / network metrics computed on the Rust side.

use std::f64::consts::PI;

/// Summary statistics for a set of weighted edges.
pub struct BasicStats {
    pub n_nodes: usize,
    pub n_edges: usize,
    pub total_length: f64,
    pub mean_length: f64,
    pub mean_out_degree: f64,
    pub self_loops: usize,
}

/// Compute basic network statistics from edge endpoints and weights.
pub fn basic_stats(from: &[usize], to: &[usize], weight: &[f64], n_nodes: usize) -> BasicStats {
    let n_edges = from.len();
    let total_length: f64 = weight.iter().sum();
    let mean_length = if n_edges > 0 { total_length / n_edges as f64 } else { 0.0 };
    let mean_out_degree = if n_nodes > 0 { n_edges as f64 / n_nodes as f64 } else { 0.0 };
    let self_loops = from.iter().zip(to).filter(|(u, v)| u == v).count();
    BasicStats { n_nodes, n_edges, total_length, mean_length, mean_out_degree, self_loops }
}

/// Shannon entropy (in nats) of edge compass bearings, binned into `num_bins`
/// equal sectors over `[0, 360)`. Higher entropy = more disordered (organic)
/// street grid; lower = more ordered (gridiron). Bearings are in degrees.
pub fn orientation_entropy(bearings: &[f64], num_bins: usize) -> f64 {
    if bearings.is_empty() || num_bins == 0 {
        return 0.0;
    }
    let mut counts = vec![0.0f64; num_bins];
    let bin_width = 360.0 / num_bins as f64;
    for &b in bearings {
        let bb = b.rem_euclid(360.0);
        let mut idx = (bb / bin_width).floor() as usize;
        if idx >= num_bins {
            idx = num_bins - 1;
        }
        counts[idx] += 1.0;
    }
    let total: f64 = counts.iter().sum();
    let mut h = 0.0;
    for c in counts {
        if c > 0.0 {
            let p = c / total;
            h -= p * p.ln();
        }
    }
    h
}

/// Initial compass bearing (degrees, clockwise from north) from point
/// `(lat1, lon1)` to `(lat2, lon2)`, both in decimal degrees.
pub fn bearing(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let (phi1, phi2) = (lat1.to_radians(), lat2.to_radians());
    let dlon = (lon2 - lon1).to_radians();
    let y = dlon.sin() * phi2.cos();
    let x = phi1.cos() * phi2.sin() - phi1.sin() * phi2.cos() * dlon.cos();
    let theta = y.atan2(x);
    (theta * 180.0 / PI).rem_euclid(360.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn perfect_grid_has_low_entropy() {
        // Two orientations only (E-W and N-S): low entropy.
        let grid = vec![0.0, 90.0, 180.0, 270.0, 0.0, 90.0, 180.0, 270.0];
        let uniform = vec![0.0, 45.0, 90.0, 135.0, 180.0, 225.0, 270.0, 315.0];
        let h_grid = orientation_entropy(&grid, 8);
        let h_uniform = orientation_entropy(&uniform, 8);
        assert!(h_grid < h_uniform);
    }

    #[test]
    fn bearing_due_east_is_90() {
        let b = bearing(0.0, 0.0, 0.0, 1.0);
        assert!((b - 90.0).abs() < 1e-6);
    }
}
